packages/plugins/jwt/src/plugin.ts

```diff
@@ -0,0 +1,242 @@
+      } catch (e) {
+        logger.error(`Failed to fetch signing key from signing provided:`, e);
+      }
+    }
+
+    return null;
+  };
+
+  return {
+    onYogaInit({ yoga }) {
+      logger = yoga.logger;
+    },
+    async onRequestParse(payload) {
+      // Try to find token in request, and reject the request if needed.
+      const token = await lookupToken(payload);
+
+      if (token == null) {
+        // If token is missing, we can reject the request based on the configuration.
+        if (normalizedOptions.reject.missingToken) {
+          throw unauthorizedError(`Unauthenticated`);
+        }
+
+        return;
+      }
+
+      try {
+        // Decode the token first, in order to get the key id to use.
+        let decodedToken: Jwt | null;
+        try {
+          decodedToken = jsonwebtoken.decode(token, { complete: true });
+        } catch (e) {
+          logger.warn(`Failed to decode JWT authentication token: `, e);
+          throw badRequestError(`Invalid authentication token provided`);
+        }
+
+        if (!decodedToken) {
+          logger.warn(
+            `Failed to extract payload from incoming token, please make sure the token is a valid JWT.`,
+          );
+
+          throw badRequestError(`Invalid authentication token provided`);
+        }
+
+        // Fetch the signing key based on the key id.
+        const signingKey = await getSigningKey(decodedToken?.header.kid);
+
+        if (!signingKey) {
+          logger.warn(
+            `Signing key is not available for the key id: ${decodedToken?.header.kid}. Please make sure signing key providers are configured correctly.`,
+          );
+
+          throw Error(`Authentication is not available at the moment.`);
+        }
+
+        // Verify the token with the signing key.
+        const verified = await verify(
+          logger,
+          token,
+          signingKey,
+          normalizedOptions.tokenVerification,
+        );
+
+        if (!verified) {
+          throw unauthorizedError(`Unauthenticated`);
+        }
+
+        if (verified) {
+          // Link the verified payload with the request (see `onContextBuilding` for the reading part)
+          payloadByRequest.set(payload.request, verified);
+        }
+      } catch (e) {
+        // User-facing errors should be handled based on the configuration.
+        // These errors are handled based on the value of "reject.invalidToken" config.
+        if (e instanceof GraphQLError) {
+          if (normalizedOptions.reject.invalidToken) {
+            throw e;
+          }
+
+          return;
+        }
+
+        // Server/internal errors should be thrown, so they can be handled by the error handler and be masked.
+        throw e;
+      }
+    },
+    onContextBuilding({ context, extendContext }) {
+      if (context.request == null) {
+        throw new Error(
+          'Request is not available on context! Make sure you use this plugin with GraphQL Yoga.',
+        );
+      }
```

Comment:
Should we throw this error even if we `normalizedOptions.extendContextFieldName` is false ? I think we should shortcut this hook by verifying if this option is true or not from the start.

We can even conditionally set the `onContextBuilding` field based on this option, no need to have this hook running if the context extension is disable I think.

Reply:
fixed, thanks!